### PR TITLE
Fixes #36017 - Remove spacing from array type settings

### PR DESCRIPTION
**Pull request by girijaasoni**
…ts to make it into a valid format

### CODE REVIEW

**Comment by stejskalleos** [comment]
ok to test

**Review by Ron-Lavi — COMMENTED** [review]
Thanks @girijaasoni !
the value which is stored in the DB still has spaces,
can we also trim the spaces from it?

Screenshot from 2023-02-08 14-59-18

**Review by Ron-Lavi — COMMENTED** [review]
We should consider trimming the values from the frontend, e.g:
https://github.com/theforeman/foreman/blob/662c4a3a3a1749846f5fc6c114dd3363c7a3bcf1/webpack/assets/javascripts/react_app/components/SettingUpdateModal/components/SettingForm/SettingForm.js#L26

Though to better protect the DB also from the API point, maybe we can use this library: https://github.com/holli/auto_strip_attributes

this can be a solution for more scenarios across Foreman and plugins...
interested in what @nofaralfasi and @theforeman/packaging have to say about it.

File: app/validators/array_type_validator.rb [hunk 1]
```diff
@@ -1,5 +1,5 @@
 class ArrayTypeValidator < ActiveModel::EachValidator
   def validate_each(record, attribute, value)
-    record.errors.add attribute, _("must be comma separated") unless Array(value).map { |item| item.to_s.match /\A\S+\z/ }.all?
+    record.errors.add attribute, _("must be comma separated") unless Array(value).map { |item| item.to_s.strip.match /\A\S+\z/ }.all?
```
Comment (ekohl) [inline app/validators/array_type_validator.rb]: Unrelated to this PR, but I think `.all?` accepts a block and this could be optimized:
```suggestion
    record.errors.add attribute, _("must be comma separated") unless Array(value).all? { |item| item.to_s.strip.match?(/\A\S+\z/) }
```

Note I didn't test this and it's probably better to do in a separate PR

**Review by ekohl — CHANGES_REQUESTED** [review]
I'm with @Ron-Lavi that I'd avoid changing the validators, but rather make sure the input is clean. As @Ron-Lavi pointed out, you will end up with more complex values in the DB which may confuse other code later on and create subtle bugs.

**Review by ekohl — CHANGES_REQUESTED** [review]
To be explicit: the changes to the validators should be reverted. They should not accept any whitespace.

File: app/validators/array_type_validator.rb (hunk 1 above)
Comment (girijaasoni) [reply app/validators/array_type_validator.rb]: > To be explicit: the changes to the validators should be reverted. They should not accept any whitespace.

@ekohl Are you suggesting that we shall apply the change somewhere else or we can modify the Regex for the same validation?

Comment (ekohl) [reply app/validators/array_type_validator.rb]: I'd apply the change elsewhere. Normally the flow is:

* User input
* Apply normalization
* Apply validators
* Process data

Processing data can mean storing it in the DB, but you want it to be as clean as possible. The main reason is that if you can assume that data in the database doesn't have whitespace around the value, you can keep the code that uses data from the database simpler. Or put in another way: if you must remember to always call `.strip` on a value, you can easily introduce bugs. That's more important than the potential performance difference, but that's another factor.

An example:
```ruby
if model.value == 'value'
  # Do X
else
  # Do Y
end
```

This will work most of the time, but if the user (accidentally) has spaces in the value it can introduce a bug.

Looking back at the flow, this means we want to clean up the data before processing it. A validator is then only supposed to validate. It never changes data. This means normalization can happen in the UI, before it's submitted, or in the controller before it's passed to the validators. But the validator must remain strict.

Comment (nofaralfasi) [reply app/validators/array_type_validator.rb]: Thanks a lot for the clarification Ewoud!!

**Comment by nofaralfasi** [comment]
@girijaasoni Some tests are failing.
And can you also update the commit message to reflect your new changes?

**Comment by Ron-Lavi** [comment]
Hey @girijaasoni, can you update this PR?

**Comment by girijaasoni** [comment]
> Hey @girijaasoni, can you update this PR?

A few tests are failing so working on it

File: app/models/setting.rb [hunk 2]
```diff
@@ -268,4 +270,10 @@ def refresh_registry_value
       definition.value_from_db = value
     end
   end
+
+  def remove_whitespaces
+    self[:value] = value.each { |a| a.strip! if a.respond_to? :strip! }
```
Comment (evgeni) [inline app/models/setting.rb:274]: this gets executed for *any* kind of setting, but I think it needs to only happen for settings of type "array", as otherwise `.each` is not defined for those

File: app/views/unattended/provisioning_templates/discovery/debian_kexec.erb [hunk 3]
```diff
@@ -40,6 +40,6 @@ require:
 {
 "kernel": "<%= @kernel_uri %>",
 "initram": "<%= @initrd_uri %>",
-"append": "url=<%= foreman_url('provision') + "&static=yes" %> interface=<%= mac %> netcfg/get_ipaddress=<%= ip %> netcfg/get_netmask=<%= mask %> netcfg/get_gateway=<%= gw %> netcfg/get_nameservers=<%= dns %> netcfg/disable_dhcp=true netcfg/get_hostname=<%= @host.name %> BOOTIF=<%= bootif %> <%= options.compact.join(' ') %>",
+"append": "url=<%= foreman_url('provision', { static: 'yes' })%> interface=<%= mac %> netcfg/get_ipaddress=<%= ip %> netcfg/get_netmask=<%= mask %> netcfg/get_gateway=<%= gw %> netcfg/get_nameservers=<%= dns %> netcfg/disable_dhcp=true netcfg/get_hostname=<%= @host.name %> BOOTIF=<%= bootif %> <%= options.compact.join(' ') %>",
```
Comment (ekohl) [inline app/views/unattended/provisioning_templates/discovery/debian_kexec.erb]: This looks like an unrelated change. Your commit has 3d600a5631927f0c9b485cd0527cc6febf0356ba as a parent while 3305096ae310c9851fc2e17591f9a2579d80db8d is the current latest in develop, which should include the changes. I suspect this is something that went wrong in a rebase.

Please rebase on the latest develop.

**Review by ekohl — CHANGES_REQUESTED** [review]

File: app/views/unattended/provisioning_templates/discovery/debian_kexec.erb (hunk 3 above)
Comment (ekohl) [reply app/views/unattended/provisioning_templates/discovery/debian_kexec.erb]: Looks like that didn't make a change. In https://community.theforeman.org/t/git-remote-fetch-pull-and-rebase/29949 I've written up some tips on rebasing.

Comment (nofaralfasi) [reply app/views/unattended/provisioning_templates/discovery/debian_kexec.erb]: Thanks!! I say let's add it here - https://github.com/theforeman/foreman/blob/develop/developer_docs/foreman_dev_setup.asciidoc :)

Comment (ekohl) [reply app/views/unattended/provisioning_templates/discovery/debian_kexec.erb]: You mean a link? Because including the content there is wrong. There are many places which aren't Foreman, but are just plain git. Smart Proxy or foreman-documentation for example.

Comment (nofaralfasi) [reply app/views/unattended/provisioning_templates/discovery/debian_kexec.erb]: I meant to add it as a link to the document, if think it doesn't belong there we can skip it.

File: app/models/setting.rb [hunk 4]
```diff
@@ -5,6 +5,7 @@ class Setting < ApplicationRecord
   extend FriendlyId
   friendly_id :name
   include ActiveModel::Validations
+  include ActiveModel::Validations::Callbacks
```
Comment (nofaralfasi) [inline app/models/setting.rb]:
```suggestion
```
I believe we can safely remove it since the code seems to be working correctly without it. Additionally, we don't have it in other places where we use the `before_validation` callback.

File: app/models/setting.rb [hunk 5]
```diff
@@ -268,4 +270,8 @@ def refresh_registry_value
       definition.value_from_db = value
     end
   end
+
+  def remove_whitespaces
+    self[:value] = value.each { |a| a.strip! if a.respond_to? :strip! }
```
Comment (nofaralfasi) [inline app/models/setting.rb:274]: I suggest using `is_a?(String)` to check if `a` is a string, which is a more precise check than `respond_to?(:strip!)`. This avoids calling `strip!` on non-string objects, which could raise an error.

```suggestion
    self[:value] = value.each { |a| a.is_a?(String) ? a.strip! : a }
```

**Review by nofaralfasi — CHANGES_REQUESTED** [review]

**Comment by Ron-Lavi** [comment]
Hey @girijaasoni, what is the status here? :)

File: test/unit/oidc_jwt_validate_test.rb [hunk 6]
```diff
@@ -44,13 +44,13 @@ def setup
     end
 
     test 'if audience is not valid' do
-      Setting['oidc_audience'] = "no-client"
+      Setting['oidc_audience'] = ["no-client"]
```
Comment (Ron-Lavi) [inline test/unit/oidc_jwt_validate_test.rb:47]: Nice catch, it was indeed moved to setting_type of array in https://github.com/theforeman/foreman/blob/7907787d7039894eba4a4e63fb85cb8a294bb0e0/db/migrate/20200326164755_change_oidc_audience_setting_type.rb#L6

**Review by Ron-Lavi — DISMISSED** [review]
Thanks @girijaasoni, I tested and it works as expected now 👍🏼

**Comment by girijaasoni** [comment]
@ekohl , could you please review the updated changes?

**Review by ekohl — DISMISSED** [review]
Changes look good to me. It would be great to have a test, but I don't know where would be a great place.

File: test/models/setting_test.rb [hunk 7]
```diff
@@ -174,6 +174,13 @@ def test_settings_should_save_strings
     assert_equal "must be comma separated", setting.errors[:value].first
   end
 
+  test "trusted_hosts should accept values with spaces" do
+    assert Setting.create(name: 'trusted_hosts', value: [])
+    setting = Setting.find_by_name("trusted_hosts")
+    setting.value = ["localhost ", "  remotehost"]
+    assert setting.save
```
Comment (stejskalleos) [inline test/models/setting_test.rb]:
```ruby
setting = Setting.create(name: 'trusted_hosts', value: ["localhost ", "  remotehost"])
assert_equal ["localhost", "remotehost"], setting.value
```

File: test/models/setting_test.rb [hunk 8]
```diff
@@ -174,6 +174,13 @@ def test_settings_should_save_strings
     assert_equal "must be comma separated", setting.errors[:value].first
   end
 
+  test "trusted_hosts should accept values with spaces" do
```
Comment (stejskalleos) [inline test/models/setting_test.rb]: `trusted_hosts should trim values with spaces`

Comment (ekohl) [reply test/models/setting_test.rb]: https://www.betterspecs.org/#should says you should make it:
```suggestion
  test "trusted_hosts accepts values with spaces" do
```

Granted, that's for RSpec, but I like the convention.

Comment (girijaasoni) [reply test/models/setting_test.rb]: Thanks @ekohl , updated it

File: test/models/setting_test.rb (hunk 7 above)
Comment (girijaasoni) [reply test/models/setting_test.rb]: Thank you @stejskalleos

**Review by stejskalleos — DISMISSED** [review]
:green_apple: **LGTM**

Failed tests are not related, but let's wait with the merge until they are fixed.

**Comment by girijaasoni** [comment]
> green_apple **LGTM**
> 
> Failed tests are not related, but let's wait with the merge until they are fixed.

The tests pass now!

**Review by nofaralfasi — APPROVED** [review]

**Review by stejskalleos — APPROVED** [review]
:green_apple: **LGTM**

Works as expected.
Thanks @girijaasoni!
